joblib/test/test_memory.py

```diff
@@ -687,31 +694,121 @@ def test__get_cache_items_to_delete(tmpdir):
 
 def test_memory_reduce_size(tmpdir):
     memory, _, _ = _setup_toy_cache(tmpdir)
-    cachedir = memory.cachedir
-    ref_cache_items = _get_cache_items(cachedir)
+    ref_cache_items = memory.store.get_cache_items()
 
     # By default memory.bytes_limit is None and reduce_size is a noop
     memory.reduce_size()
-    cache_items = _get_cache_items(cachedir)
+    cache_items = memory.store.get_cache_items()
     assert sorted(ref_cache_items) == sorted(cache_items)
 
     # No cache items deleted if bytes_limit greater than the size of
     # the cache
     memory.bytes_limit = '1M'
     memory.reduce_size()
-    cache_items = _get_cache_items(cachedir)
+    cache_items = memory.store.get_cache_items()
     assert sorted(ref_cache_items) == sorted(cache_items)
 
     # bytes_limit is set so that only two cache items are kept
     memory.bytes_limit = '3K'
     memory.reduce_size()
-    cache_items = _get_cache_items(cachedir)
+    cache_items = memory.store.get_cache_items()
     assert set.issubset(set(cache_items), set(ref_cache_items))
     assert len(cache_items) == 2
 
     # bytes_limit set so that no cache item is kept
     bytes_limit_too_small = 500
     memory.bytes_limit = bytes_limit_too_small
     memory.reduce_size()
-    cache_items = _get_cache_items(cachedir)
+    cache_items = memory.store.get_cache_items()
     assert cache_items == []
+
+
+def test_cachedir_deprecation_warning(tmpdir):
+    # verify the right deprecation warnings are raised when using cachedir
+    # option instead of new location parameter.
+    with warns(None) as w:
+        memory = Memory(location=tmpdir.strpath, cachedir=tmpdir, verbose=0)
+        assert memory.store.cachedir.startswith(tmpdir.strpath)
+
+    assert len(w) == 1
+    assert "You set both location and cachedir options" in str(w[-1].message)
+
+    with warns(None) as w:
+        memory = Memory(cachedir=tmpdir.strpath, verbose=0)
+        assert memory.store.cachedir.startswith(tmpdir.strpath)
+
+    assert len(w) == 1
+    assert "cachedir option is deprecated since version" in str(w[-1].message)
+
+
+class IncompleteStoreBackend(StoreBackendBase):
+    """This backend cannot be instanciated and should raise a TypeError."""
+    pass
+
+
+class DummyStoreBackend(StoreBackendBase):
+    """A dummy store backend that does nothing."""
+
+    def create_location(self, location):
+        """Create location on store."""
+        "Does nothing"
+
+    def exists(self, obj):
+        """Check if an object exists in the store"""
+        return False
+
+    def clear_location(self, obj):
+        """Clear object on store"""
+        "Does nothing"
+
+    def get_cache_items(self):
+        """Returns the whole list of items available in cache."""
+        return []
+
+    def configure(self, location, *args, **kwargs):
+        """Configure the store"""
+        "Does nothing"
+
+
+def test_register_invalid_store_backends():
```

Comment:
I think this one is a good candidate for pytest @parametrize and split in 2 separate test functions.